### PR TITLE
Make use of safe log for residuals calculation

### LINKED ISSUE #1160: Add `LogSafe` function for residuals calculation.

### PR DESCRIPTION
**Pull request by PavelBal**
Fixes #1160

### CODE REVIEW

File: R/data-combined.R [hunk 1]
```diff
@@ -324,6 +324,10 @@ DataCombined <- R6::R6Class(
     #'   scale factors to apply to raw values. The default offset is `0`, while
     #'   default scale factor is `1`, i.e., the data will not be modified. If a
     #'   list is specified, it should be the same length as `forNames` argument.
+    #' @param reset IF `TRUE`, only data transformations that are specified will
```
Comment (PavelBal) [inline R/data-combined.R:327]: @svavil Documentation of the new argument was missing.

File: R/ospsuite-env.R [hunk 2]
```diff
@@ -54,6 +54,9 @@ ospsuiteEnv$isPKSimLoaded <- FALSE
 # NetTask `DimensionTask` cached for performance benefits. Created the first time it is requested.
 ospsuiteEnv$dimensionTask <- NULL
 
+# Small value added to zero when calculating log
+ospsuiteEnv$LOG_SAFE_EPSILON <- 1e-20
```
Comment (PavelBal) [inline R/ospsuite-env.R:58]: Using the same value as in Core

File: R/plot-observed-vs-simulated.R [hunk 3]
```diff
@@ -115,6 +115,16 @@ plotObservedVsSimulated <- function(dataCombined,
     return(NULL)
   }
 
+  # In logarithmic scale, if any of the values are `0`, plotting will fail.
+  #
+  # To avoid this, just remove rows where any of the quantities are `0`s.
+  if (obsVsPredPlotConfiguration$yAxis$scale %in% c(tlf::Scaling$log, tlf::Scaling$ln)) {
```
Comment (PavelBal) [inline R/plot-observed-vs-simulated.R:121]: This was previously done during residuals calculation, but is only relevant for plotting.

File: R/utilities-data-combined.R [hunk 4]
```diff
@@ -214,12 +214,18 @@ calculateResiduals <- function(dataCombined,
...
     )$y
   )
-
   # Residual computation will depend on the scaling.
   if (scaling %in% c(tlf::Scaling$lin, tlf::Scaling$identity)) {
     pairedData <- dplyr::mutate(pairedData, residualValues = yValuesSimulated - yValuesObserved)
   } else {
-    pairedData <- dplyr::mutate(pairedData, residualValues = log(yValuesSimulated) - log(yValuesObserved))
+    # Epsilon for safe log calculation should be converted to the units of the values
+    epsilon <- toUnit(
```
Comment (PavelBal) [inline R/utilities-data-combined.R:222]: The epsilon should only be valid for values in base unit. So have to convert it here to the actually used unit.

File: R/utilities-data-combined.R [hunk 5]
```diff
@@ -214,12 +214,18 @@ calculateResiduals <- function(dataCombined,
...
+      quantityOrDimension = pairedData$yDimension[[1]],
+      values = ospsuiteEnv$LOG_SAFE_EPSILON,
+      targetUnit = pairedData$yUnit[[1]],
+      molWeight = 1
+    )
+    pairedData <- dplyr::mutate(pairedData, residualValues = .log_safe(yValuesSimulated, epsilon = epsilon) - .log_safe(yValuesObserved, epsilon = epsilon))
```
Comment (PavelBal) [inline R/utilities-data-combined.R:228]: Using log_safe now.

File: R/utilities-data-combined.R [hunk 6]
```diff
@@ -229,40 +235,30 @@ calculateResiduals <- function(dataCombined,
...
-  x <- ospsuite.utils::toMissingOfType(x, type = "double")
-
-  if (is.na(x)) {
-    return(NA_real_)
-  } else if (x < epsilon) {
-    return(log(epsilon, base = base))
-  } else {
-    return(log(x, base = base))
-  }
+.log_safe <- function(x, base = 10, epsilon = ospsuiteEnv$LOG_SAFE_EPSILON) {
```
Comment (PavelBal) [inline R/utilities-data-combined.R:249]: Also now using log_10 and not ln to produce the same results as pksim/mobi

File: R/utilities-data-combined.R [hunk 7]
```diff
@@ -229,40 +235,30 @@ calculateResiduals <- function(dataCombined,
...
+  x <- sapply(X = x, \(element){
```
Comment (PavelBal) [inline R/utilities-data-combined.R:250]: Also vectorized the function.

File: tests/testthat/test-utilities-data-combined.R [hunk 8]
```diff
@@ -6,15 +6,15 @@ simFilePath <- system.file("extdata", "Aciclovir.pkml", package = "ospsuite")
 sim <- loadSimulation(simFilePath)
 simData <- withr::with_tempdir({
   df <- dplyr::tibble(
-    IndividualId = c(0, 0, 0),
-    `Time [min]` = c(0, 2, 4),
-    `Organism|PeripheralVenousBlood|Aciclovir|Plasma (Peripheral Venous Blood) [µmol/l]` = c(0, 4, 8)
+    IndividualId = c(0, 0, 0, 0),
```
Comment (PavelBal) [inline tests/testthat/test-utilities-data-combined.R:9]: Extended the tests to account for the new behavior for log_safe(0)

File: R/plot-observed-vs-simulated.R (hunk 3 above)
Comment (msevestre) [inline R/plot-observed-vs-simulated.R:121]: can it be NOT in log or ln?

File: R/ospsuite-env.R (hunk 2 above)
Comment (msevestre) [reply R/ospsuite-env.R:58]: The plan it to use the one from cor eventually right? (the logsafe function)

File: R/utilities-data-combined.R (hunk 4 above)
Comment (msevestre) [reply R/utilities-data-combined.R:222]: hum..good point

File: R/utilities-data-combined.R [hunk 9]
```diff
@@ -229,40 +235,30 @@ calculateResiduals <- function(dataCombined,
...
+    element <- ospsuite.utils::toMissingOfType(element, type = "double")
+    if (is.na(element)) {
+      return(NA_real_)
+    } else if (element < epsilon) {
```
Comment (msevestre) [inline R/utilities-data-combined.R:254]: you don't need all the else here (just FYI no need to chnagE)

**Review by msevestre — COMMENTED** [review]
Looks good to me
I just have the feeling that I saw somewhere a method asking to use the method from .net. Is this temporary or what's the difference?

File: R/plot-observed-vs-simulated.R (hunk 3 above)
Comment (PavelBal) [reply R/plot-observed-vs-simulated.R:121]: It can be linear `tlf::Scaling$lin`.

**Comment by PavelBal** [comment]
> The plan it to use the one from cor eventually right? (the logsafe function)

I don't think we should make .NET calls for such simple operation. If logsafe were more complex, yes, but here I would prefer to have this function implemented in R.

**Comment by PavelBal** [comment]
> I just have the feeling that I saw somewhere a method asking to use the method from .net. Is this temporary or what's the difference?

We discussed about using residuals calculation from Core. Still under discussion, I just wanted to maker sure that the current implementation is robust.

**Comment by PavelBal** [comment]
@msevestre The build is passing, so feel free to merge if you don't have any other remarks.

**Pull request merged**